slatedb/src/manifest/store.rs

```diff
@@ -80,115 +73,100 @@ impl FenceableManifest {
             manifest_update_timeout,
             system_clock,
+            |m: &Manifest| m.writer_epoch,
+            |m: &mut Manifest, e: u64| m.writer_epoch = e,
         )
-        .await
+        .await?;
+        Ok(Self { inner: fr })
     }
 
     pub(crate) async fn init_compactor(
         stored_manifest: StoredManifest,
         manifest_update_timeout: Duration,
         system_clock: Arc<dyn SystemClock>,
     ) -> Result<Self, SlateDBError> {
-        Self::init(
-            stored_manifest,
-            |m| m.compactor_epoch,
-            |m, e| m.compactor_epoch = e,
+        let fr = FenceableRecord::init(
+            stored_manifest.inner,
             manifest_update_timeout,
             system_clock,
+            |m: &Manifest| m.compactor_epoch,
+            |m: &mut Manifest, e: u64| m.compactor_epoch = e,
         )
-        .await
-    }
-
-    async fn init(
-        mut stored_manifest: StoredManifest,
-        stored_epoch: fn(&Manifest) -> u64,
-        set_epoch: fn(&mut DirtyManifest, u64),
-        manifest_update_timeout: Duration,
-        system_clock: Arc<dyn SystemClock>,
-    ) -> Result<Self, SlateDBError> {
-        utils::timeout(
-            system_clock,
-            manifest_update_timeout,
-            "manifest update",
-            async {
-                loop {
-                    let local_epoch = stored_epoch(&stored_manifest.manifest) + 1;
-                    let mut manifest = stored_manifest.prepare_dirty();
-                    set_epoch(&mut manifest, local_epoch);
-                    match stored_manifest.update_manifest(manifest).await {
-                        Err(ManifestVersionExists) => {
-                            // The manifest may have been updated by a reader, or
-                            // we may have gotten this error after successfully updating
-                            // if we failed to get the response. Either way, refresh
-                            // the manifest and try the bump again.
-                            stored_manifest.refresh().await?;
-                            continue;
-                        }
-                        Err(err) => return Err(err),
-                        Ok(()) => {
-                            return Ok(Self {
-                                stored_manifest,
-                                local_epoch,
-                                stored_epoch,
-                            })
-                        }
-                    }
-                }
-            },
-        )
-        .await
-        .map_err(|_| SlateDBError::Timeout {
-            op: "manifest update",
-            backoff: Duration::from_secs(1),
-        })
+        .await?;
+        Ok(Self { inner: fr })
     }
 
     pub(crate) async fn refresh(&mut self) -> Result<(), SlateDBError> {
-        self.stored_manifest.refresh().await?;
-        self.check_epoch()
+        self.inner.refresh().await?;
+        Ok(())
```

Comment:
Nit: Does this work?

```suggestion
        self.inner.refresh().await
```